cmd/permissions-checker/permissionschecks.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os/user"
+	"regexp"
+	"strings"
+	"syscall"
+)
+
+// FindRecursive iterate in a directory (path) showing permissions in a recursive way
+/* path  - where to look for files and directories
+ * level - used to print the output in a hierarchical way
+ */
+func FindRecursive(path, level string) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// In verbose mode we are printing the output in a hierarchical way
+	if Verbose {
+		level = level + "  "
+	}
+
+	for _, f := range files {
+		name := f.Name()
+		relativePath := strings.TrimPrefix(strings.Join([]string{path, name}, "/"), Directory)
+		fullPath := strings.Join([]string{path, name}, "/")
+
+		match, _ := regexp.MatchString(Exclude, relativePath)
+		if match {
+			if Verbose {
+				fmt.Printf("\x1b[33m%sExcluding %s\n\x1b[0m", level, name)
+			}
+		} else {
+			mode := f.Mode()
+			currentOwner, _ := user.LookupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Uid))
+			currentGroup, _ := user.LookupGroupId(fmt.Sprint(f.Sys().(*syscall.Stat_t).Gid))
+			currentPerm := mode.Perm().String()
+			hasCorrectOwner := checkOwner(currentOwner.Username, DefaultOwner)
+			hasCorrectGroup := checkOwner(currentGroup.Name, DefaultGroup)
+
+			if mode.IsRegular() && (f.Name()[0:1] != "." || ShowHidden) {
+				hasCorrectPermissions := checkPermissions(currentPerm, DefaultFilePerm)
+				printOutput(level, "f", name, fullPath, currentPerm, DefaultFilePerm, currentOwner.Username, currentGroup.Name, hasCorrectPermissions, hasCorrectOwner, hasCorrectGroup)
+			} else if mode.IsDir() && (f.Name()[0:1] != "." || ShowHidden) {
+				hasCorrectPermissions := checkPermissions(currentPerm, DefaultDirPerm)
+				printOutput(level, "d", name, fullPath, currentPerm, DefaultDirPerm, currentOwner.Username, currentGroup.Name, hasCorrectPermissions, hasCorrectOwner, hasCorrectGroup)
+				FindRecursive(fullPath, level)
+			}
+		}
+	}
+}
+
+// checkPermissions return true if the permissions are correct (false in other case)
+/* currentPermissions [string] - permissions that the file or directory has
+ * defaultPermissions [string] - permissions that should have
+ */
+func checkPermissions(currentPermissions, defaultPermissions string) bool {
```

Comment:
Can you add an example of what permissions are? Is it `755`? How does using `strings.Contains` compare it?
Consider normalizing the user-provided to the expected format before checking

Reply:
`currentPermissions` is the value obtained from `mode.Perm().String()` and it is something like `-rwxrwxr-x` (note that the first character is `-`)
`defaultPermissions` is the value passed by the user (or the default one) and it can contain the first character `-` or not, that is the reason I am using `strings.Contains`

I can remove the first character in `currentPermissions` to unify it, what do you think?